flocker/control/httpapi.py

```diff
@@ -54,6 +69,106 @@ def version(self):
+    @user_documentation("""
+        Create a new dataset.
+        """, examples=[
+            u"create dataset",
+            u"create dataset with dataset_id",
+            u"create dataset with duplicate dataset_id",
+            u"create dataset with maximum_size",
+            u"create dataset with metadata",
+        ])
+    @structured(
+        inputSchema={'$ref': '/v1/endpoints.json#/definitions/datasets'},
+        outputSchema={'$ref': '/v1/endpoints.json#/definitions/datasets'},
+        schema_store=SCHEMAS
+    )
+    def create_dataset(self, primary, dataset_id=None, maximum_size=None, metadata=None):
+        """
+        Create a new dataset in the cluster configuration.
+
+        :param unicode primary: The address of the node on which the primary
+            manifestation of the dataset will be created.
+
+        :param unicode dataset_id: A unique identifier to assign to the
+            dataset.  This is typically a UUID.  If no value is given, one will
+            be generated and returned in the response.  This is not for easy
+            human use.  For human-friendly identifiers, use items in
+            ``metadata``.
+
+        :param int maximum_size: The maximum number of bytes the dataset will
+            be capable of storing.  This may be optional or required depending
+            on the dataset backend.
+
+        :param dict metadata: A small collection of unicode key/value pairs to
+            associate with the dataset.  These items are not interpreted.  They
+            are only stored and made available for later retrieval.  Use this
+            for things like human-friendly dataset naming, ownership
+            information, etc.
+
+        :return: A ``dict`` describing the dataset which has been added to the
+            cluster configuration or giving error information if this is not
+            possible.
+        """
+        if dataset_id is None:
+            dataset_id = unicode(uuid4())
+
+        if metadata is None:
+            metadata = {}
+
+        # Use persistence_service to get a Deployment for the cluster
+        # configuration.
+        deployment = self.persistence_service.get()
+        for node in deployment.nodes:
+            for manifestation in node.manifestations():
+                if manifestation.dataset.dataset_id == dataset_id:
+                    raise DATASET_ID_COLLISION
+
+        # XXX Check cluster state to determine if the given primary node
+        # actually exists.  If not, raise PRIMARY_NODE_NOT_FOUND.
+
+        dataset = Dataset(
+            dataset_id=dataset_id,
+            maximum_size=maximum_size,
+            metadata=pmap(metadata)
+        )
+        manifestation = Manifestation(dataset=dataset, primary=True)
+
+        primary_nodes = list(
+            node for node in deployment.nodes if primary == node.hostname
+        )
+        if len(primary_nodes) == 0:
+            primary_node = Node(hostname=primary)
```

Comment:
:exclamation: It seem like the comment above suggests this should raise `PRIMARY_NODE_NOT_FOUND`.